domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -205,10 +223,65 @@ class ExplorationProgressController @Inject constructor(
     }
   }
 
+  fun submitUnrevealedHintIsVisible(
```

Comment:
Honestly, I'm starting to think these changes are more complicated than just moving the hint/solution logic to this controller.

I'm partly wondering if we could just move HintHandler to this class and, when a hint/solution availability changes, notify that the ephemeral state DataProvider has changed (and ensure the necessary hint/solution data is included within EphemeralState). At that point, all we need to track is which hints are revealed & whether the solution is revealed. All the extra piping to keep the UI & domain layers in sync would become much simpler.

Beyond that, we also can probably then simplify what needs to be stored to retain hint/solution state. One HintHandler is needed per session (& we should be resetting it across session states) which means we don't need to keep track of hint sequence number or tracked wrong answer count since these can be reasonably reset upon resuming a checkpoint. Ditto I think for isHintVisibleInLatestState (which would mean only HelpIndex actually needs to be stored within the checkpoint).

Reply:
@BenHenning Moving hint handler to the domain layer will simplify the flow but it might create clashes with @yashraj-01's project. Because I think in one of the PRs he has to move the handler out of the StatePlayerRecyclerViewAssemebler as a separate class to show hints in developer options.

@BenHenning @yashraj-01 WDYT?

Reply:
@MaskedCarrot I'm not sure if this is a good idea. I also thought of moving the HintHandler to the domain layer but it interacts with the UI in some cases. For example, it calls the listeners while showing a hint. Also, in my PR #3613, we are putting the HintHandler modules in the FragmentComponent so, I'm not sure how it will affect your new implementation or vice-versa.

@BenHenning any thoughts on this?

Reply:
Good point--this is an unexpected clash between the two projects. Unfortunately, the complexity we're seeing here trying to make this all work might force our hand toward trying to move it to the domain layer.

My thinking here is that we change the flow to the following: StateFragment reactively shows hints/solution entirely based on EphemeralState (mitigating the need for the listener at all--instead, we would notify that the DataProvider has changed in that same location when the class is moved to the domain layer).

The module changes should be workable: we can just use ApplicationComponent, instead, since this becomes an application-level thing instead of a fragment-level thing. I suspect it won't change too much for your project @yashraj-01, but it will definitely change some of it. I also suspect that it will probably be easier to base your work off of @MaskedCarrot's, but we probably should have a separate, preliminary PR that moves the handler to the domain layer so that it's not coupled with the rest of this resume work & so that your work isn't blocked for as long.

Would this be workable?

Reply:
I think moving the hint handler to the domain layer might get complicated. We have lifecycle aware timers that the hint handler uses to show hints. Since these timers are being observed as live data, they will have to remain in the app layer. Also, the delay for which the timer runs depends upon the current state, i.e. it is 60 seconds for the first hint, 30 for others and 10 seconds if the user submits an answer. Since the delay is variable and not fixed, we will have to compute that in the app layer. 

So I think we will have to distribute the hint handler between the app and domain layers, which might complicate the solution.

@BenHenning  WDYT?

Reply:
Actually, I was able to move the hint handler to the domain layer, and I think it does simplify the implementation of both the hint handler and the resuming mechanism. 
There are a couple of edge cases left that I still need to take care of. I will probably send a PR in a few hours or by tomorrow.

Reply:
Created a new PR (##3659) to move the hint handler to the domain module.